mcs/class/System/System.Net.Sockets/SafeSocketHandle.cs

```diff
@@ -55,20 +56,24 @@ protected override bool ReleaseHandle ()
 						break;
 					}
 
-					/*
-					 * This method can be called by the DangerousRelease inside RegisterForBlockingSyscall
-					 * When this happens blocking_threads contains the current thread.
-					 * We can safely close the socket and throw SocketException in RegisterForBlockingSyscall
-					 * before the blocking system call.
-					 */
 					lock (blocking_threads) {
+						/*
+						* This method can be called by the DangerousRelease inside RegisterForBlockingSyscall
+						* When this happens blocking_threads contains the current thread.
+						* We can safely close the socket and throw SocketException in RegisterForBlockingSyscall
+						* before the blocking system call.
+						*/
 						if (blocking_threads.Count == 1 && blocking_threads[0] == Thread.CurrentThread)
 							break;
-					}
 
-					AbortRegisteredThreads ();
-					// Sleep so other threads can resume
-					Thread.Sleep (1);
+						// abort registered threads
+						foreach (var t in blocking_threads)
+							Socket.cancel_blocking_socket_operation (t);
+
+						// Sleep so other threads can resume
+						in_cleanup = true;
+						Monitor.Wait (blocking_threads, 1000);
```

Comment:
I don't think this will work unless the Monitor.Wait is outside the lock.

If we do Monitor.Wait inside the lock then UnRegisterForBlockingSyscall () won't be able to do Monitor.Pulse because it is also inside a lock.